feat(protocol-engine): Support loading pipettes and labware from JSON protocols

SyntaxColoring opened the pull request:
# Overview

When a JSON protocol is uploaded through the experimental HTTP API, to be run through protocol engine, implement the missing parts of loading the pipettes and labware that the JSON protocol uses. Closes #7946.

# Changelog

Main changes:

* Changed `CommandTranslator` to take in an entire protocol at once, instead of going command-by-command.
  * Changed calling code and its tests accordingly.
* Gave `CommandTranslator` the ability to translate a JSON protocol's choices of pipettes and labware into Protocol Engine `LoadPipetteRequest`s, `AddLabwareDefinitionRequest`s, and `LoadLabwareRequest`s.

Also:

* Clarified that `LoadPipetteRequest.pipetteName` is the pipette *load* name, as opposed to the model name or something.
* Left a maybe unnecessary comment in `opentrons/protocols/models/__init.py__` describing when (I think) it should re-export things from its child submodules. I had gone down an unproductive path adding re-exports for everything I could possibly need from that tree.

# Deferred work

* I made a mess of the tests, which I will clean up in a follow-up PR. See the `todo` comment at the top of `test_command_translator.py` for the fix approach that @mcous and I agreed on.
  * For alternative approaches considered, see these todo comments from an earlier version of this PR, which I've since removed: https://github.com/Opentrons/opentrons/pull/7950#discussion_r653875070, https://github.com/Opentrons/opentrons/pull/7950#discussion_r653879429.
  * Ticket for follow-up: #7958
* There's some confusing "id"/"URI"/"UUID" terminology between PD and PE that we should reconcile at some pont.

# Other stuff discussed in in-person reviews

* @sanni-t and I considered renaming `CommandTranslator` to `ProtocolTranslator`, since it now takes in and translates an entire JSON protocol at once, rather than going command-by-command. This could have been necessary if we chose a testing approach where command translation is a separate class from protocol translation, but we didn't, so this renaming doesn't feel worthwhile to me anymore.

# Review requests

I would focus on the production code, since we know the tests to be messy and we already have a good path in mind to fix them.

# Risk assessment

Low. Only used in our experimental Protocol Engine work, behind a feature flag.

SyntaxColoring committed:
Make CommandTranslator take in the whole protocol at once.

SyntaxColoring committed:
WIP: Emit AddLabwareDefinitionRequests and LoadLabwareRequests.

SyntaxColoring committed:
Fix up whitespace.

SyntaxColoring committed:
Clean up imports.

SyntaxColoring committed:
Assert loading a labware comes after adding its respective definition.

SyntaxColoring committed:
Fix whitespace.

SyntaxColoring committed:
Clarify that LoadPipetteRequest.pipetteName is the load name.

SyntaxColoring committed:
Emit LoadPipetteRequests.

SyntaxColoring committed:
Avoid casts where unnecessary.

SyntaxColoring committed:
Translate liquid-handling commands. Also, general cleanup.

SyntaxColoring committed:
Simplify imports.

SyntaxColoring committed:
Use "pe" instead of "protocol_engine".

mcous commented:
Just did some real world testing, and it's looking like this is fundamentally working! We've got a [little issue with the command state](https://github.com/Opentrons/opentrons/pull/7882#discussion_r653782940), but otherwise this looks like it's doing the thing 🎉

SyntaxColoring committed:
Resolve todo.

As we've just implemented things in pipette and labware translation, yes, these IDs are the same.

SyntaxColoring committed:
Clean up test imports.

SyntaxColoring committed:
Elaborate on testing todo comment.

SyntaxColoring committed:
Autoformat tests with black.

SyntaxColoring committed:
Minimize, and document intent of, modules/__init__.py re-exports.

* Revert some thoughtless re-exports I made earlier in this PR.
* Say what I think actually should be re-exported here, so I don't make the same mistake again.

SyntaxColoring committed:
Clarify todo comment for duplicated JSON protocol fixture.

SyntaxColoring committed:
Delete stray todo.

SyntaxColoring commented on api/src/opentrons/protocols/runner/json_proto/command_translator.py:
```diff
@@ -22,10 +20,32 @@ def __init__(self) -> None:
-    def translate(self, command: models.AllCommands) -> ReturnType:
-        """
-        Translate a PD/JSON protocol to Protocol Engine commands.
+    def translate(self, protocol: models.JsonProtocol) -> PECommands:
```
Let's return something like `List[pe.commands.CommandRequestType]`, for specificity. (Following the general practice that functions should be typed to return things that are specific and to take in things that are generic.)

SyntaxColoring commented on api/tests/opentrons/conftest.py:
```diff
@@ -565,7 +565,9 @@ def minimal_labware_def():
 def minimal_labware_def2():
```
Should also be typed to return dict.

SyntaxColoring commented on api/tests/opentrons/protocols/runnner/json_proto/test_command_translator.py:
```diff
@@ -1,92 +1,254 @@
+from opentrons.protocols.runner.json_proto.command_translator import CommandTranslator
+
+
+# todo(mm, 2021-06-17):
+#
+# This JSON protocol fixture is copy-pasted from an opentrons.file_runner fixture.
+# Either lift the fixture somewhere common to both of these modules, or move one of
+# the modules next to the other so they can benefit from the same fixtures.
```
Or don't use one big fixture at all. Have helper functions to construct minimal protocols with given labware, commands, etc.

SyntaxColoring replied:
☝️ Going with that approach and removing this todo. See PR description.

SyntaxColoring commented on api/tests/opentrons/protocols/runnner/json_proto/test_command_translator.py:
```diff
@@ -1,92 +1,254 @@
+    assert expected_request in result
+
+
+# todo(mm, 2021-06-17):
+#
+# These tests use the internal implementation detail _translate_command.
+#
+# Maybe _translate_command(), _translate_add_labware_definition(), and
+# _translate_load_labware() should be static methods in classes (or a single class)
+# separate from CommandTranslator. Then these methods could be public again.
+#
+# Otherwise, we should delete these tests in favor of treating CommandTranslator
+# as a black box.
```
In @mcous's experience, this will start feeling like busy work. e.g. "if you receive an `AspirateCommand`, make sure you call `translate_aspirate_command()`..."

SyntaxColoring replied:
Going with a different approach and removing this todo. See PR description.

SyntaxColoring committed:
Type minimal_labware_def2() as returning a dict.

SyntaxColoring committed:
Type all methods as returning lists, not iterables.

SyntaxColoring committed:
Update test todo comment to reflect in-person review discussion.

mcous commented on api/tests/opentrons/protocols/runnner/json_proto/test_command_translator.py:
```diff
@@ -1,92 +1,254 @@
+import typing
```
Fun game: look at this file's full path and tell me what the typo is

SyntaxColoring replied:
Fixing this typo in https://github.com/Opentrons/opentrons/pull/7957. ⌨️

mcous commented on api/src/opentrons/protocols/models/__init__.py:
```diff
@@ -1,3 +1,11 @@
+# Convenience re-exports of models that are especially common or important.
+# More detailed sub-models are always available through the underlying
+# submodules.
+#
+# If re-exporting something, its name should still make sense when it's separatred
+# from the name of its parent submodule. e.g. re-exporting models.json_protocol.Labware
+# as models.Labware could be confusing.
```
I see, this is referring to the `from .json_protocol import Model as JsonProtocol` re-export business? Agree that should be avoided

SyntaxColoring replied:
Not totally sure we're talking about the same thing.

I had gone down an unproductive path of adding re-exports for everything I could possibly need from `json_protocol` and `labware_definitions` to this `__init__.py`. I tried re-exporting `json_protocol.Labware`, `json_protocol.AllCommands`, and so on.

Somewhere in the middle of this PR, I realized that this would cause confusing naming problems. For example, `models.Labware` sounds like it's about Opentrons labware definitions in general, but as a re-export of `json_protocols.Labware`, it's actually a very specific JSON protocol thing.

So this comment is meant to say, "I know we're re-exporting some things here at the top level, but don't get tempted into doing this for absolutely everything. Here's where it does and doesn't make sense."

`json_protocol.Model` is funky but I don't think the concern is specific to it.

mcous commented on api/src/opentrons/protocols/runner/json_proto/command_translator.py:
```diff
@@ -1,18 +1,16 @@
-    PickUpTipRequest, DropTipRequest
-)
-from opentrons.protocols.models import json_protocol as models
+import opentrons.types
```
I find `import foo.bar` to be very confusing; it's was non-obvious to me that you then proceed to use `foo.bar` in the code.

Given that we use exactly two structures from here, I would've written this as:

```py
from opentrons.types import DeckSlotName, MountType
```

Mostly because I'm lazy and it's less typing everywhere, and doesn't feel to me like it's a loss in understanding because the name is consistent everywhere and most of us are working in editors / IDEs with type info + click-to-follow

mcous commented on api/src/opentrons/protocols/runner/json_proto/command_translator.py:
```diff
@@ -22,10 +20,32 @@ def __init__(self) -> None:
-        Translate a PD/JSON protocol to Protocol Engine commands.
+    def translate(self, protocol: models.JsonProtocol) -> PECommands:
+        """Return all Protocol Engine commands required to run the given protocol."""
+        result: List[pe.commands.CommandRequestType] = []
```
`pe.commands.WhateverThing` makes me feel a little like we're locking into the internal shape of the `protocol_engine` module. It sort of makes me think that this translator belongs in the `protocol_engine` module itself (perhaps in `clients` or something

SyntaxColoring replied:
Because we mention `commands` over and over every time we type `pe.commands.WhateverThing`, as opposed to once at the top of the file like `from protocol_engine.commands import ...`?

That's true.

I think in this particular case, it gets resolved if we do `from protocol_engine import commands as pe_commands`, as you suggested and as [I've just written about in Slack](https://opentrons.slack.com/archives/G01LDD5U3C6/p1623965369017500).

This also brings up a second discussion of whether submodule names are / can be / should be part of the whole collective package's stable interface. I'll write about this separately.

mcous reviewed (approved):
Some discussion points, but covered most of this on chat. Got a few discussion points below, but otherwise this looks to be fundamentally working and the TODOs check out

sanni-t commented on api/tests/opentrons/protocols/runnner/json_proto/test_command_translator.py:
```diff
@@ -1,92 +1,254 @@
+# * Making every test here use that helper function with its own hard-coded choice of
+#   commands. (So we don't have to pull expected details out of a fixture to assert
+#   against.)
+# * Consistently testing CommandTranslator as a black box here.
```
I thought testing CommandTranslator as black box meant having one big integrated test that used a complete protocol fixture and tested for presence of all the commands. Since that's not what the rest of the points suggest, can you elaborate what this last point means?

SyntaxColoring replied:
Yeah, sorry, I wrote this unclearly.

It becomes something roughly like this:

```python
def test_aspirate_translation():
    json_aspirate_command = models.json_protocol.Aspirate(...)

    # This is the trick. Each test still uses a full JSON protocol, but it's tailored
    # individually for each test to be as minimal as possible, through this
    # _create_protocol() test helper function.
    #
    # Maybe we also need to add a pipette and labware here, idk.
    json_protocol = _create_protocol(commands=[json_aspirate_command])

    result = CommandTranslator().translate(json_protocol)

    assert protocol_engine.commands.AspirateRequest(...) in result
```

sanni-t replied:
Thanks for clarifying

SyntaxColoring commented:
Merging this now to make it easier to test running protocols on robots (yay!). But feel free to keep leaving comments here (and maybe ping me) for stuff to fix in a follow-up.
